metadata/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+        bytes.extend(storage.prefix.as_bytes());
+        for entry in storage.entries.iter() {
+            bytes.extend(entry.name.as_bytes());
+            entry.modifier.encode_to(&mut bytes);
+            match &entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    bytes.extend(get_type_hash(registry, ty.id(), &mut visited_ids));
+                }
+                StorageEntryType::Map {
+                    hashers,
+                    key,
+                    value,
+                } => {
+                    hashers.encode_to(&mut bytes);
+                    bytes.extend(get_type_hash(registry, key.id(), &mut visited_ids));
+                    bytes.extend(get_type_hash(registry, value.id(), &mut visited_ids));
+                }
+            }
+            bytes.extend(&entry.default);
+        }
+    }
+
+    hash(&bytes)
+}
+
+/// Obtain the hash representation of a `frame_metadata::RuntimeMetadataLastVersion`.
+pub fn get_metadata_hash(metadata: &RuntimeMetadataLastVersion) -> MetadataHashDetails {
+    // Collect all pairs of (pallet name, pallet hash).
+    let mut pallets: Vec<(String, [u8; 32])> = metadata
+        .pallets
+        .iter()
+        .map(|pallet| {
+            let name = pallet.name.clone();
+            let hash = get_pallet_hash(&metadata.types, pallet);
+            (name, hash)
+        })
+        .collect();
+    // Sort by pallet name to create a deterministic representation of the underlying metadata.
+    pallets.sort_by_key(|key| key.1);
+
+    // Note: pallet name is excluded from hashing.
+    // Each pallet has a hash of 32 bytes, and the vector is extended with
+    // extrinsic hash and metadata ty hash (2 * 32).
+    let mut bytes = Vec::with_capacity(pallets.len() * 32 + 64);
+    for (_, hash) in pallets.iter() {
+        bytes.extend(hash)
+    }
+
+    bytes.extend(get_extrinsic_hash(&metadata.types, &metadata.extrinsic));
+
+    let mut visited_ids = HashSet::<u32>::new();
+    bytes.extend(get_type_hash(
+        &metadata.types,
+        metadata.ty.id(),
+        &mut visited_ids,
+    ));
+
+    let hash = hash(&bytes);
+
+    MetadataHashDetails {
+        metadata_hash: hash,
+        pallet_hashes: pallets.into_iter().collect(),
+    }
+}
+
+/// Metadata hasher internal cache.
+#[derive(Clone, Debug)]
+pub struct MetadataHashDetails {
+    /// Full metadata hash.
+    pub metadata_hash: [u8; 32],
+    /// Pallet hashes resulted from metadata hashing.
+    pub pallet_hashes: HashMap<String, [u8; 32]>,
+}
+
+impl MetadataHashDetails {
+    /// Creates an empty `MetadataHasherCache`.
+    pub fn new() -> MetadataHashDetails {
+        Self {
+            metadata_hash: [0; 32],
+            pallet_hashes: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MetadataHashDetails {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        get_metadata_hash,
+        get_pallet_hash,
+    };
```

Comment:
The convention is normally just to `use super::*` to bring everything in the module being tested into scope :)